### ISSUE TITLE
Revise pmix_proc_t to reduce memory footprint

### ISSUE THREAD

**Issue by rhc54** [pmix/pmix-standard#366]
### Overview

Consider replacing the fixed-size character array in `pmix_proc_t` with a `char*` pointer

### Motivation

The memory footprint of a PMIx implementation can become quite large when supporting an application at scale. In terms of scaling on a per-process basis, the size of the footprint is negatively impacted by inclusion of the `pmix_nspace_t` field in the `pmix_proc_t` identifier associated with each process as that field is defined as a fixed-size character of 256 bytes.

### Discussion Items

The Standard currently defines the size of `pmix_nspace_t` as 256 bytes (`PMIX_MAX_NSLEN` of 255 plus one location for the `NULL` terminator). While there have been instances where the job-level identifier of an application reached such sizes, these instances have been rare - the far more common case is for the identifier to be a string representation of a number, perhaps with some decoration that brings the total length to the 10-20 character range. The Standard does provide for a reduction in the size by specifying a _minimum_ size of 64 bytes, but that needs to be removed in order to stabilize the ABI. Thus, implementations will be forced to settle on the 256-byte number.

Reducing the memory footprint of implementations would be aided by changing the definition of `pmix_nspace_t` to a `char*` pointer, thus minimizing the storage required by the `pmix_proc_t`. The rationale behind the initial definition was to ensure that a shared memory representation of an array of `pmix_proc_t` would have a fixed step size between elements. However, this could be maintained by ensuring that the string in the `pmix_proc_t` not be inlined - i.e., that the `char*` pointer not point to a string contained within the `pmix_proc_t` itself. This is an implementation-level decision, but is noted here to explain why the described change could be made without violating the initial rationale.

The question for discussion therefore is: does the Standard want to revise the definition of `pmix_nspace_t`? If so, then what should be done - reduce the size of the fixed array, change it to a `char*` pointer, or...?

**Comment by jjhursey** [comment]
I think we should entertain moving away from the fixed-length array for the namespace and instead move to something that lends itself to more flexible representations. This will be an ABI break, but something to do now that we are first defining the ABI.

I have a couple of options to consider.

## Current

Currently, we have (taking these from OpenPMIx `pmix_common.h` for simplicity):
```
#define PMIX_MAX_NSLEN     255

typedef uint32_t pmix_rank_t;
typedef char pmix_nspace_t[PMIX_MAX_NSLEN+1];

typedef struct pmix_proc {
    pmix_nspace_t nspace;
    pmix_rank_t rank;
} pmix_proc_t;
```

Assorted notes:
 * (-) For every proc structure being used by the client we have `255` characters, often not used.
 * (-) Comparing namespaces requires a string operation.
 * (+) We don't have to explicitly free the namespace in the proc structure.
 * (+) Not changing it doesn't break any existing PMIx consumers

## Option 1: Replace fixed array with string pointer

```
typedef uint32_t pmix_rank_t;
typedef char* pmix_nspace_t;

typedef struct pmix_proc {
    pmix_nspace_t nspace;
    pmix_rank_t rank;
} pmix_proc_t;
```

Assorted notes:
 * (+) Only the space required for the namespace string is allocated by the PMIx implementation
 * (+) Namespace can be any length
 * (-) Who frees the memory? Users would likely be in charge of free'ing the namespace even though PMIx allocated it.

## Option 2: Replace fixed array with an integer

```
typedef uint32_t pmix_rank_t;
typedef uint32_t pmix_nspace_t;

typedef struct pmix_proc {
    pmix_nspace_t nspace;
    pmix_rank_t rank;
} pmix_proc_t;
```

Assorted notes:
 * (+) Often the namespace is already a string representation of an integer. Where it isn't then a reference table can be used when the "true" namespace is needed to be known.
 * (+) Don't have to worry about free'ing memory
 * (+) No string comparison operations on the namespace.

The namespace is a good first-level grouping of processes in the job. Often we compare namespaces against each other (via `strncmp`) and not necessarily need to deconstruct them (for those systems that encode multiple values into the namespace - such as batch job and job step).

We could provide a function that converts the integer PMIx namespace into the string representation of the namespace if the client needs to use it to talk directly to the RM (without PMIx). But I don't think that's something folks should be doing in general - meaning that's a misuse of the namespace (right?).

**Comment by rhc54** [comment]
As per the comment [here](https://github.com/pmix/pmix-standard/issues/367#issuecomment-994912086), I recommend we close this issue and leave `pmix_proc_t` as-is.

**Comment by rhc54** [comment]
I've been coding a prototype of this change in OpenPMIx, and have found the following:

* there is one API signature that has a problem. `PMIx_Spawn` includes a `pmix_nspace_t` parameter that the library fills with the namespace of the spawned job. The current definition guarantees that this parameter is backed by memory. However, the revised definition leaves it unbacked by memory - and the address of space allocated inside the library cannot be returned. Short of changing the signature, the only solution is to revise the text in the Standard to require that the user provide allocated space with minimum size of `PMIX_MAX_NSLEN+1`.
* I had to change the `pmix_value_t` definition - the current data union contains a `pmix_nspace_t*` that must be changed to `pmix_nspace_t`. This actually helps to reduce the size of the `pmix_value_t` structure.
* anyone who uses the macros is otherwise protected from the change. Modifications to the macros aren't terribly intrusive and fully "hide" the change in nspace type
* anyone who did not use the macros is likely to encounter trouble. The current definition guarantees a static allocation of memory for the namespace's string. Changing `pmix_nspace_t` to a `char*` removes that guarantee, thereby leading to potential segfaults - e.g., doing a `strncpy` into a `pmix_nspace_t` will now fail.

Compilers sometimes help to catch these problems by generating "not initialized" warnings, but this does not provide complete protection. In fact, most of the changes I had to make to OpenPMIx stemmed from directly using `strncpy` instead of the PMIx macros.

Bottom line: we can make this work, BUT it will force source-code modifications by current users, particularly by those who did not use the macros (e.g., people who `dlopen` the library instead of linking against it) or use `PMIx_Spawn`. This may be alleviated once the `pmix-abi` headers are released, and is a one-time change. However, there may be issues when building across versions that would require version protection, which could get messy.

**Comment by jjhursey** [comment]
Note from IA WG meeting on Jan. 24, 2022 (revisited on March 14, 2022):
 * Inside the PMIx library the string associated with the namespace can be handled in a more flexible manner than how it is exposed to the user, primarily, via the `pmix_proc_t`.
 * If the user uses the macros to create/manipulate/destroy the `pmix_proc_t` then the PMIx library could hide the data structure used for the `pmix_nspace_t` (e.g., `char *` vs `char [255]`) from the user. However, there are cases where users are probably already accessing the namespace in other ways.
 * As such, any change to the `pmix_nspace_t` would be a big disruptive change in a number of APIs.
 * **Decision:** It was the decision of the working group to leave the `pmix_nspace_t` as it is today. If additional uses cases are brought forward then we will certainly reconsider. It was the feeling of the working group that most of the scenarios where it might be a problem (e.g., proc table) can be solved in other ways (e.g., new attributes and associated data structures).

Per the notes above I'm going to close this ticket. We can reopen it if further discussion is desired.
